[autoscaler] Replace cluster yaml validation with json schema validator

eisber opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

- improved YAML parse error message
- Support for IntelliSense in VSCode and other IDEs

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22248/
Test FAILed.

richardliaw commented on python/setup.py:
```diff
@@ -174,6 +174,8 @@ def find_version(*filepath):
     "pytest",
     "pyyaml",
+    "jsonschema",
+    "importlib_resources",
```
BTW, porting the comment from the Autoscaler PR over here: do we need importlib_resources? This is possible to do with just `f.read(os.path.join(ray.__file__, "ray-schema.json"))` right?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22553/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22554/
Test PASSed.

richardliaw commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -764,61 +764,13 @@ def kill_workers(self):
 
-    check_required(config, schema)
-    check_extraneous(config, schema)
+    schema = f.read(os.path.join(ray.autoscaler.__file__, "ray-schema.json")) 
```
sorry, this needs to be:
```suggestion
    with open(os.path.join(ray.autoscaler.__file__, "ray-schema.json")) as f:
        schema = f.read() 
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22567/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22566/
Test PASSed.

eisber commented:
> I just tried this out; I actually am not sure if this works.
> 
> I added an extra key `hi` in the yaml and ran the autoscaler, but it didn't catch this extraneous key:
> 
> ```
> # An unique identifier for the head node and workers of this cluster.
> cluster_name: minimal
> hi: 1
> # The maximum number of workers nodes to launch in addition to the head
> # node. This takes precedence over min_workers. min_workers default to 0.
> min_workers: 0
> max_workers: 0
> 
> provider:
>     type: aws
>     region: us-west-2
>     availability_zone: us-west-2b
> 
> # How Ray will authenticate with newly launched nodes.
> auth:
>     ssh_user: ubuntu
> 
> head_node:
>     InstanceType: m4.10xlarge
>     ImageId: ami-3b6bce43  # Amazon Deep Learning AMI (Ubuntu)
> 
> setup_commands:
>     - error me
>     #     - echo 'export PATH="$HOME/anaconda3/envs/tensorflow_p36/bin:$PATH"' >> ~/.bashrc
>     #     - source activate tensorflow_p27 && pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.4.0-cp27-cp27mu-manylinux1_x86_64.whl
> 
> # # Command to start ray on the head node. You don't need to change this.
> head_start_ray_commands:
>     - ray stop
>     - ray start --head --redis-port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml
> ```
> 
> Is this intended behaviour?

I added additional restrictions. Once I can build on my box again, I'll add a unit test

pcmoritz commented:
This is great! Can you also add a unit test to make sure the existing yaml files we ship with the autoscaler pass the validations? In particular, in https://github.com/ray-project/ray/tree/master/python/ray/autoscaler we have

```
aws/development-example.yaml
aws/example-full.yaml
aws/example-gpu-docker.yaml
aws/example-minimal.yaml

gcp/example-full.yaml
gcp/example-gpu-docker.yaml
gcp/example-minimal.yaml

kubernetes/example-full.yaml
kubernetes/example-minimal.yaml

local/development-example.yaml
local/example-full.yaml
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22589/
Test FAILed.

eisber commented:
> > I just tried this out; I actually am not sure if this works.
> > I added an extra key `hi` in the yaml and ran the autoscaler, but it didn't catch this extraneous key:
> > ```
> > # An unique identifier for the head node and workers of this cluster.
> > cluster_name: minimal
> > hi: 1
> > # The maximum number of workers nodes to launch in addition to the head
> > # node. This takes precedence over min_workers. min_workers default to 0.
> > min_workers: 0
> > max_workers: 0
> > 
> > provider:
> >     type: aws
> >     region: us-west-2
> >     availability_zone: us-west-2b
> > 
> > # How Ray will authenticate with newly launched nodes.
> > auth:
> >     ssh_user: ubuntu
> > 
> > head_node:
> >     InstanceType: m4.10xlarge
> >     ImageId: ami-3b6bce43  # Amazon Deep Learning AMI (Ubuntu)
> > 
> > setup_commands:
> >     - error me
> >     #     - echo 'export PATH="$HOME/anaconda3/envs/tensorflow_p36/bin:$PATH"' >> ~/.bashrc
> >     #     - source activate tensorflow_p27 && pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.4.0-cp27-cp27mu-manylinux1_x86_64.whl
> > 
> > # # Command to start ray on the head node. You don't need to change this.
> > head_start_ray_commands:
> >     - ray stop
> >     - ray start --head --redis-port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml
> > ```
> > 
> > 
> > Is this intended behaviour?
> 
> I added additional restrictions. Once I can build on my box again, I'll add a unit test

added the unit tests

eisber commented:
> This is great! Can you also add a unit test to make sure the existing yaml files we ship with the autoscaler pass the validations? In particular, in https://github.com/ray-project/ray/tree/master/python/ray/autoscaler we have
> 
> ```
> aws/development-example.yaml
> aws/example-full.yaml
> aws/example-gpu-docker.yaml
> aws/example-minimal.yaml
> 
> gcp/example-full.yaml
> gcp/example-gpu-docker.yaml
> gcp/example-minimal.yaml
> 
> kubernetes/example-full.yaml
> kubernetes/example-minimal.yaml
> 
> local/development-example.yaml
> local/example-full.yaml
> ```

added those as well

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22621/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22664/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22728/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22744/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22770/
Test FAILed.

richardliaw commented:
btw; it looks like the merge base was messed up. Would it be possible to reopen a new PR and ping me when the diff looks right? 

Sorry for the delay.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22901/
Test FAILed.

richardliaw commented on ci/travis/install-dependencies.sh:
```diff
@@ -53,7 +53,7 @@ if [[ "$PYTHON" == "3.6" ]] && [[ "$platform" == "linux" ]]; then
     feather-format lxml openpyxl xlrd py-spy pytest-timeout networkx tabulate aiohttp \
     uvicorn dataclasses pygments werkzeug kubernetes flask grpcio pytest-sugar pytest-rerunfailures pytest-asyncio \
-    blist scikit-learn numba
+    blist scikit-learn numba parameterized pytest
```
pytest should already be installed

eisber replied:
I was a bit surprised that you install pytest as part of ray and not introduce ray[test]

richardliaw replied:
maybe let's do this in a separate PR, since I think it'll affect some documentation

richardliaw commented on python/ray/tests/test_autoscaler_yaml.py:
```diff
@@ -25,6 +27,31 @@ def testValidateDefaultConfig(self):
+            jsonschema.validate(config, json.loads(schema))
+            self.fail("Expected validation to fail")
+        except:
+            pass
```
Instead of this, can you make a helper method `_test_invalid_config`, and then create separate tests for each set of yamls?

`testInvalidConfigKubernetes` (here, call `_test_invalid_config` on "autoscaler/kubernetes/example-full.yaml", "autoscaler/kubernetes/example-minimal.yaml"),

`testInvalidConfigLocal` (here, call `_test_invalid_config` on "autoscaler/local/example-full.yaml", "autoscaler/local/example-minimal.yaml"),

`testInvalidConfigGCP` (here, call `_test_invalid_config` on "autoscaler/gcp/example-full.yaml", "autoscaler/gcp/example-minimal.yaml"),

I would like to avoid introducing a new dependency (`parameterized`). After all, the test output doesn't actually seem that easy to decipher  `https://pypi.org/project/parameterized/`

eisber replied:
sure can change this. I hoped that in case of failure it's a bit easier to figure as it lists the config file.

richardliaw replied:
Yeah... i'm not sure (from the examples, it doesn't seem to be the case)

richardliaw commented on python/setup.py:
```diff
@@ -76,6 +76,7 @@
     "dashboard": [],
     "serve": ["uvicorn", "pygments", "werkzeug", "flask", "pandas", "blist"],
     "tune": ["tabulate", "tensorboardX"],
+    "test": ["pytest", "parameterized"]
```
if we remove parameterized, can we revert this?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22945/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22946/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22977/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22984/
Test PASSed.
